wasmcloud-test-util/src/provider_test.rs

```diff
@@ -2,58 +2,163 @@
 pub type TomlMap = std::collections::BTreeMap<String, toml::Value>;
 pub type JsonMap = serde_json::Map<String, serde_json::Value>;
 
-const DEFAULT_START_DELAY_SEC: u64 = 1;
-const DEFAULT_RPC_TIMEOUT_MILLIS: u64 = 2000;
+const DEFAULT_START_DELAY: Duration = Duration::from_secs(1);
+const DEFAULT_RPC_TIMEOUT: Duration = Duration::from_secs(2);
 const DEFAULT_NATS_URL: &str = "127.0.0.1:4222";
 // use a unique lattice prefix to separate test traffic
 const TEST_LATTICE_PREFIX: &str = "TEST";
 const TEST_HOST_ID: &str = "_TEST_";
 
-static ONCE: OnceCell<Provider> = OnceCell::const_new();
-pub type TestFunc = fn() -> BoxFuture<'static, RpcResult<()>>;
+static PROVIDERS: OnceCell<Mutex<HashMap<Config, Provider>>> = OnceCell::const_new();
 
-fn to_value_map(data: &toml::map::Map<String, TomlValue>) -> RpcResult<SimpleValueMap> {
-    let mut map = SimpleValueMap::default();
-    // copy simple values into the map
-    for (k, v) in data.iter() {
-        match v {
-            TomlValue::Integer(_) | TomlValue::Float(_) | TomlValue::Boolean(_) => {
-                map.insert(k.clone(), v.to_string());
-            }
-            // string is handled separately because 'to_string()' adds quotes
-            TomlValue::String(s) => {
-                map.insert(k.clone(), s.clone());
-            }
-            // intentionally omitted
-            TomlValue::Array(_) | TomlValue::Table(_) | TomlValue::Datetime(_) => {}
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct LogLevel(pub log::Level);
+
+impl Deref for LogLevel {
+    type Target = log::Level;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl FromStr for LogLevel {
+    type Err = <log::Level as FromStr>::Err;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        s.parse::<log::Level>().map(Self)
+    }
+}
+
+impl<'de> Deserialize<'de> for LogLevel {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        String::deserialize(deserializer)?
+            .parse()
+            .map_err(de::Error::custom)
+    }
+}
+
+impl From<log::Level> for LogLevel {
+    fn from(level: log::Level) -> Self {
+        Self(level)
+    }
+}
+
+impl From<LogLevel> for log::Level {
+    fn from(LogLevel(level): LogLevel) -> Self {
+        level
+    }
+}
+
+/// Provider test config
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq)]
+pub struct Config {
+    /// Whether to enable Rust backtrace
+    pub backtrace: bool,
+    /// Contract ID
+    pub contract_id: String,
+    /// Lattice RPC prefix
+    pub lattice_rpc_prefix: String,
+    pub link_delay: Duration,
+    pub link_name: String,
+    /// [`LogLevel`] for capability provider set using "RUST_LOG" environment variable, defaults to [LogLevel::Info]
+    pub log_level: LogLevel,
+    /// URL at which NATS is accessible
+    pub nats_url: String,
+    /// URL at which Redis is accessible
+    pub redis_url: String,
```

Comment:
Although`redis_url` was in the sample config (commented out),  I don't think it should be part of the `Config` struct since most tests won't need it. If a test does need it, it fits better into the `values` map.

`nats_url` should be here because nearly ever test _does_ need that